lib/src/main/kotlin/com/swmansion/starknet/data/serializers/MerkleNodePolymorphicSerializer.kt

```diff
@@ -0,0 +1,21 @@
+import com.swmansion.starknet.data.types.*
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.json.JsonContentPolymorphicSerializer
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.jsonObject
+
+internal object MerkleNodePolymorphicSerializer :
+    JsonContentPolymorphicSerializer<NodeHashToNodeMappingItem.MerkleNode>(NodeHashToNodeMappingItem.MerkleNode::class) {
+    override fun selectDeserializer(element: JsonElement): DeserializationStrategy<NodeHashToNodeMappingItem.MerkleNode> {
+        val jsonElement = element.jsonObject
+        val binaryNodeKeys = NodeHashToNodeMappingItem.BinaryNode.serializer().descriptor.elementNames.toSet()
+        val edgeNodeKeys = NodeHashToNodeMappingItem.EdgeNode.serializer().descriptor.elementNames.toSet()
+
+        return when (jsonElement.keys) {
+            binaryNodeKeys -> NodeHashToNodeMappingItem.BinaryNode.serializer()
+            edgeNodeKeys -> NodeHashToNodeMappingItem.EdgeNode.serializer()
+            else -> throw IllegalArgumentException("Invalid MerkleNode JSON object: $jsonElement")
+        }
+    }
+}
```

Comment:
note: We could potentially give an error with more meaningful message (saying which fields are missing and if it's binary/edge node). However, in case of json with keys that belong both to binary and edge node
e.g.
```
{
    "left": ...,
    "path": ...
}
```
we can't clearly decide which type it is intended to be (binary/edge). I'm not fully sure if such logic is worth introducing. 

A simple improvement would be to display required keys for binary/edge nodes.